Framework/Nexus/inc/MantidNexus/NexusIOHelper.h

```diff
@@ -186,19 +245,38 @@ std::vector<T> readNexusVector(::NeXus::File &file,
 template <typename T>
 std::vector<T> readNexusSlab(::NeXus::File &file, const std::string &entry,
                              const std::vector<int64_t> &start,
-                             const std::vector<int64_t> &size) {
-  auto info_and_close =
+                             const std::vector<int64_t> &size,
+                             const bool allow_downcasting = false) {
+  const auto info_and_close =
       checkIfOpenAndGetInfo(file, std::move(std::move(entry)));
   RUN_NEXUSIOHELPER_FUNCTION((info_and_close.first).type, readNexusAnySlab,
-                             file, start, size, info_and_close.second);
+                             file, start, size, info_and_close.second,
+                             allow_downcasting);
+}
+
+/** Opens the data group if needed, finds the data type, and calls
+ * readNexusAnySlab via the RUN_NEXUSIOHELPER_FUNCTION macro.
+ * The provided output buffer is filled.
+ */
+template <typename T>
+void readNexusSlab(std::vector<T> &out, ::NeXus::File &file,
+                   const std::string &entry, const std::vector<int64_t> &start,
+                   const std::vector<int64_t> &size,
+                   const bool allow_downcasting = false) {
+  const auto info_and_close =
+      checkIfOpenAndGetInfo(file, std::move(std::move(entry)));
+  RUN_NEXUSIOHELPER_FUNCTION((info_and_close.first).type, readNexusAnySlab, out,
+                             file, start, size, info_and_close.second,
+                             allow_downcasting);
 }
 
 template <typename T>
-T readNexusValue(::NeXus::File &file, const std::string &entry = "") {
-  auto info_and_close =
+T readNexusValue(::NeXus::File &file, const std::string &entry = "",
+                 const bool allow_downcasting = false) {
```

Comment:
Do we need the downcasting flag to be a runtime flag, or could this be a compile time (template argument)?

Reply:
Nice suggestion. I think it can be a compile-time argument.